AVRO-2389: CSharp serializers for plain C# objects

pa009fa opened the pull request.

pa009fa commented:
Have to fix project files for .net framework

Fokko commented:
I've restarted the job, it looks like a network issue.

@blachniet What do you think? My C# is a bit rusty.

blachniet commented:
@pa009fa, this is going to be a great addition to the C# library! Thank you for sharing!

I also wanted to let you know that I've pulled the code and started to review the changes. Hopefully we can get this in soon!

pa009fa commented:
Happy to help.

Is there any documentation I need to work on? (I still need to do the /// comments).

From: Brian Lachniet <notifications@github.com>
Reply-To: apache/avro <reply@reply.github.com>
Date: Friday, May 17, 2019 at 11:27 AM
To: apache/avro <avro@noreply.github.com>
Cc: Patrick Farry <Patrick.Farry@pb.com>, Mention <mention@noreply.github.com>
Subject: Re: [apache/avro] [AVRO-2389] CSharp serializers for plain C# objects (#521)

CAUTION: THIS EMAIL IS FROM AN EXTERNAL SOURCE. Internet links, office documents or other attachments may contain viruses. Do not click on a link, open or enable any file unless you trust the sender.

@pa009fa<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpa009fa&data=02%7C01%7CPatrick.Farry%40pb.com%7Ca4d797a7333f4593957508d6daf55a4c%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636937144647597070&sdata=izK9c%2FuAq6OKwbFNcHKHhQmpj%2Fhp22zrOX8Vv8%2FWnZQ%3D&reserved=0>, this is going to be a great addition to the C# library! Thank you for sharing!

I also wanted to let you know that I've pulled the code and started to review the changes. Hopefully we can get this in soon!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%3Femail_source%3Dnotifications%26email_token%3DAIDVNMV3RMDEKHUE35OEXALPV32J5A5CNFSM4HMLJ4Z2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVVQNCI%23issuecomment-493553289&data=02%7C01%7CPatrick.Farry%40pb.com%7Ca4d797a7333f4593957508d6daf55a4c%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636937144647607079&sdata=HFgCdtHt%2Bgrpb%2FH47U1NoI2h6KW%2BriHh%2BBUgCXLyXy4%3D&reserved=0>, or mute the thread<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAIDVNMXBITZ33RKXUSCVYPLPV32J5ANCNFSM4HMLJ4ZQ&data=02%7C01%7CPatrick.Farry%40pb.com%7Ca4d797a7333f4593957508d6daf55a4c%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636937144647607079&sdata=9%2F%2B514O1ox7CojqZzpJMOziGlzsfjLKNQRE26%2BJ6uLo%3D&reserved=0>.

blachniet commented on lang/csharp/src/apache/main/POCO/AvroAttribute.cs:
```diff
@@ -0,0 +1,52 @@
+
+ using System;
+
+namespace Avro.POCO
```
The Java API has a similar feature called "Reflect". Should we stick with the "POCO" term to make it easier for C# devs to grok, or call it "Reflect" to match other Avro clients?

blachniet commented on lang/csharp/src/apache/main/POCO/POCOWriter.cs:
```diff
@@ -0,0 +1,250 @@
+            if (fixedrec == null)
+            {
+                throw new AvroTypeException("Fixed object is not derived from byte[]");
+            }
```
I think it would be nice to also allow derivatives of `GenericFixed` as `value` in this method. This would allow users of the library to use some of their auto-generated Specific API classes in the POCO API.

blachniet commented on lang/csharp/src/apache/main/POCO/POCOWriter.cs:
```diff
@@ -0,0 +1,250 @@
+        /// <param name="encoder">The encoder for serialization</param>
+        protected override void WriteArray(ArraySchema schema, object value, Encoder encoder)
+        {
+            var arr = value as System.Collections.IList;
```
Could we allow `IEnumerable<T>` here rather than requiring an `IList`? `DefaultSpecificWriter` also requires an `IList` at this time, but maybe we could imporve on this here. Would allowing an `IEnumerable` cause problems?

blachniet commented on lang/csharp/src/apache/main/POCO/POCOWriter.cs:
```diff
@@ -0,0 +1,250 @@
+        /// <param name="encoder">encoder to write to</param>
+        protected override void WriteMap(MapSchema schema, object value, Encoder encoder)
+        {
+            if (value == null)
```
This is a very sensible check. We should probably add this to `DefaultSpecificWriter` as well (not necessarily as part of this PR). Actually, if not for this check, this `WriteMap` method is identical to the base implementation. After running this check, why not call `base.WriteMap()`?

blachniet commented on lang/csharp/src/apache/main/POCO/POCOWriter.cs:
```diff
@@ -0,0 +1,250 @@
+        /// <param name="encoder">The encoder for serialization</param>
+        protected override void WriteUnion(UnionSchema us, object value, Encoder encoder)
+        {
+            for (int i = 0; i < us.Count; i++)
```
`WriteUnion` appears identical to the base implementation. Do we really need to override it?

pa009fa commented:
Hi Brian,

Thanks for the reply. Responses follow:


  1.  .net framework build: Need to fix this but I don’t have easy access to a windows machine. The system doesn’t build cleanly on my Mac. I figured I would put the net40 targets back in when the pull request was closer to being final (or just revert the changes). I can do this now if you think it’s better.
  2.  Field numbers: I was trying to match the behavior of the codegen classes. I agree using a field name attribute would be better – I’ll make the change.
  3.  Reflect vs. POCO. Happy with either. I guess Reflect for consistency. Will make the change.
  4.  Fixed type to support GenericFixed. Yes, can do that.
  5.  IEnumerable vs. IList: Will try it out. I think the code only ever iterates through the collection so it should work.
  6.  WriteMap. Yes Will do. I hadn’t realized the implementation didn’t change much. Thanks.
  7.  WriteUnion. Same comment as for WriteMap.

I noticed the class caching isn’t going to work when the object has an array or map at the top level. Will fix this too and add tests.

From: Brian Lachniet <notifications@github.com>
Reply-To: apache/avro <reply@reply.github.com>
Date: Tuesday, June 11, 2019 at 4:21 AM
To: apache/avro <avro@noreply.github.com>
Cc: Patrick Farry <Patrick.Farry@pb.com>, Mention <mention@noreply.github.com>
Subject: Re: [apache/avro] AVRO-2389: CSharp serializers for plain C# objects (#521)

CAUTION: THIS EMAIL IS FROM AN EXTERNAL SOURCE. Internet links, office documents or other attachments may contain viruses. Do not click on a link, open or enable any file unless you trust the sender.

@blachniet requested changes on this pull request.

25 days later... Sorry my delay in getting back. I've got some initial review feedback. I'd still like to dive into this deeper and get a better understanding of the nuts and bolts, but I wanted to share what I've got so far.

________________________________

In lang/csharp/src/apache/main/Avro.main.csproj<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r284416378&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156389378&sdata=chkdEEK2sSZaGQpv3cG5%2BNTGfDHEXaSVG%2B7IJvnC%2Blg%3D&reserved=0>:

> @@ -19,7 +19,7 @@



   <Import Project="../../../common.props" />







   <PropertyGroup>



-    <TargetFrameworks>net40;netstandard2.0</TargetFrameworks>



+    <TargetFrameworks>netstandard2.0</TargetFrameworks>



Was there something in these changes that prevented you from targeting net40?

We are currently supporting .NET Framework 4.0+ in the Avro C# libraries. Removing it here would be a breaking change. I've been toying with the idea of dumping support for anything prior to 4.6.1 (.NET Standard 2.0 supports .NET Framework 4.6.1+<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Fdotnet%2Fstandard%2Fnet-standard%23net-implementation-support&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156399387&sdata=2L6YMRiQfP%2FPvCI9LnN%2B4zPbnsjngagc68AcoAJFMCg%3D&reserved=0>) but I have not actually gotten to the point of proposing that to the Avro team yet.

If we were to drop support for .NET 4.0, I think we would probably need to wait for a major release, e.g. 1.10.x. For now, I suggest that you add net40 back in to the target frameworks.
⬇️ Suggested change

-    <TargetFrameworks>netstandard2.0</TargetFrameworks>



+    <TargetFrameworks>net40;netstandard2.0</TargetFrameworks>



________________________________

In lang/csharp/src/apache/main/POCO/AvroAttribute.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292396937&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156409395&sdata=6o12xYGph19PMIMLjhnWtADFI3u4WDWElIrdIO4x8fY%3D&reserved=0>:

> +namespace Avro.POCO



+{



+    /// <summary>



+    /// Attribute to indicate how the class is mapped to the schema. If true the mapping is by position



+    /// and the class must have field attributes. If false the class is mapped by field name. Mapping by field



+    /// position allows you to use properties who's name doesnt match the schema.



+    ///



+    /// If there is no attribute mapping by name is the default.



+    /// </summary>



+    public class AvroAttribute : Attribute



+    {



+        /// <summary>



+        /// If true the class is mapped by position.



+        /// </summary>



+        /// <value></value>



+        public bool ByPosition { get; set; }



Why map properties to a field at a particular position? This seems fragile, especially since Avro schema evolution allows for the reordering of fields in a schema, which would change their "position".

In the documentation for this class, you mention that this would be used when your C# property name cannot match your Avro field name. This is definitely a scenario we want to handle. Maybe we should take an approach similar to that of the JsonPropertyAttribute<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.newtonsoft.com%2Fjson%2Fhelp%2Fhtml%2FJsonPropertyName.htm&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156409395&sdata=nnUxPFJcAFjGmL8waqyeycwCVxYZkBUe7wFKqoTBsgM%3D&reserved=0>. In the Avro approach, the AvroFieldAttribute would store the Avro field name to map a particular property to.

For example, the schema below defines a "release-date" field, which cannot be replicated exactly in a C# property name because of the hyphen.

{



  "type": "record",



  "name": "Movie",



  "fields" : [



    {"name": "title", "type": "string"},



    {"name": "release-date", "type": "long"}



  ]



}

I imagine we could map the C# property name to the appropriate Avro field name using the AvroFieldAttribute like so:

class Movie



{



    public string Title { get; set }







    [AvroFieldAttribute("release-date")]



    public long ReleaseDate { get; set; }



}

________________________________

In lang/csharp/src/apache/main/POCO/AvroAttribute.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292399459&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156419399&sdata=roQEmVD9UueL1qMO0Zy43iQP3zfqEOWwZ0Dwuf6jYqI%3D&reserved=0>:

> + * to you under the Apache License, Version 2.0 (the



+ * "License"); you may not use this file except in compliance



+ * with the License.  You may obtain a copy of the License at



+ *



+ *     http://www.apache.org/licenses/LICENSE-2.0



+ *



+ * Unless required by applicable law or agreed to in writing, software



+ * distributed under the License is distributed on an "AS IS" BASIS,



+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.



+ * See the License for the specific language governing permissions and



+ * limitations under the License.



+ */



+



+ using System;



+



+namespace Avro.POCO



The Java API has a similar feature called "Reflect". Should we stick with the "POCO" term to make it easier for C# devs to grok, or call it "Reflect" to match other Avro clients?

________________________________

In lang/csharp/src/apache/main/POCO/POCOWriter.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292400612&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156419399&sdata=kWMXnn7DjDJQxhI0OUANZmdpTZ%2FbNfRz7CFxgxR1Wa8%3D&reserved=0>:

> +        }



+



+        /// <summary>



+        /// Validates that the record is a fixed record object and that the schema in the object is the



+        /// same as the given writer schema. Writes the given fixed record into the given encoder



+        /// </summary>



+        /// <param name="schema">writer schema</param>



+        /// <param name="value">fixed object to write</param>



+        /// <param name="encoder">encoder to write to</param>



+        protected override void WriteFixed(FixedSchema schema, object value, Encoder encoder)



+        {



+            var fixedrec = value as byte[];



+            if (fixedrec == null)



+            {



+                throw new AvroTypeException("Fixed object is not derived from byte[]");



+            }



I think it would be nice to also allow derivatives of GenericFixed as value in this method. This would allow users of the library to use some of their auto-generated Specific API classes in the POCO API.

________________________________

In lang/csharp/src/apache/main/POCO/POCOWriter.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292401231&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156429412&sdata=Gn9RttvoJY76m4onVeiDc8Z7641M21oK8JKJu7pOs%2Fo%3D&reserved=0>:

> +            }



+



+            encoder.WriteFixed(fixedrec);



+        }



+



+        /// <summary>



+        /// Serialized an array. The default implementation calls EnsureArrayObject() to ascertain that the



+        /// given value is an array. It then calls GetArrayLength() and GetArrayElement()



+        /// to access the members of the array and then serialize them.



+        /// </summary>



+        /// <param name="schema">The ArraySchema for serialization</param>



+        /// <param name="value">The value being serialized</param>



+        /// <param name="encoder">The encoder for serialization</param>



+        protected override void WriteArray(ArraySchema schema, object value, Encoder encoder)



+        {



+            var arr = value as System.Collections.IList;



Could we allow IEnumerable<T> here rather than requiring an IList? DefaultSpecificWriter also requires an IList at this time, but maybe we could imporve on this here. Would allowing an IEnumerable cause problems?

________________________________

In lang/csharp/src/apache/main/POCO/POCOWriter.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292401952&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156439416&sdata=6FKfHTBnXt1BS1QMqT3QU%2BipRcaS4FSVcdGLowIK0go%3D&reserved=0>:

> +                encoder.StartItem();



+                Write(schema.ItemSchema, arr[i], encoder);



+            }



+



+            encoder.WriteArrayEnd();



+        }



+



+        /// <summary>



+        /// Writes the given map into the given encoder.



+        /// </summary>



+        /// <param name="schema">writer schema</param>



+        /// <param name="value">map to write</param>



+        /// <param name="encoder">encoder to write to</param>



+        protected override void WriteMap(MapSchema schema, object value, Encoder encoder)



+        {



+            if (value == null)



This is a very sensible check. We should probably add this to DefaultSpecificWriter as well (not necessarily as part of this PR). Actually, if not for this check, this WriteMap method is identical to the base implementation. After running this check, why not call base.WriteMap()?

________________________________

In lang/csharp/src/apache/main/POCO/POCOWriter.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292402258&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156439416&sdata=kONy9EQwJU1cuvmFICQm%2BcFhzsKa9R7oH1q3AyAQ%2FXQ%3D&reserved=0>:

> +            }



+



+            encoder.WriteMapEnd();



+        }



+



+        /// <summary>



+        /// Resolves the given value against the given UnionSchema and serializes the object against



+        /// the resolved schema member. The default implementation of this method uses



+        /// ResolveUnion to find the member schema within the UnionSchema.



+        /// </summary>



+        /// <param name="us">The UnionSchema to resolve against</param>



+        /// <param name="value">The value to be serialized</param>



+        /// <param name="encoder">The encoder for serialization</param>



+        protected override void WriteUnion(UnionSchema us, object value, Encoder encoder)



+        {



+            for (int i = 0; i < us.Count; i++)



WriteUnion appears identical to the base implementation. Do we really need to override it?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%3Femail_source%3Dnotifications%26email_token%3DAIDVNMRN4OQ4KWJ63UTLBNTPZ6DFDA5CNFSM4HMLJ4Z2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBYX76PI%23pullrequestreview-238026557&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156449429&sdata=ezNR3l280LcFHMG%2F6G9TK97bHZDW%2BjULXcDg%2FF5ToqM%3D&reserved=0>, or mute the thread<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAIDVNMU6LNZYT23SG63NBSLPZ6DFDANCNFSM4HMLJ4ZQ&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156449429&sdata=QXU6g0bp7q1KsPNhwxJ1L0BuatLT6uP80UTwiFnpkf8%3D&reserved=0>.

pa009fa commented:
Brian,

Hope these emails are getting to you. Would you mind to reply? If I don’t hear back I’ll paste the thread into the jira ticket.

Regarding GenericFixed
The problem is the deserializer. The signature of the call we inherit from GenericDefaultReader looks like this:
        protected override object ReadFixed(object reuse, FixedSchema writerSchema, Schema readerSchema, Decoder d)
If reuse is null (probably the usual case) then we don’t know the type of the object we are trying to create. A general fix for this would be to extend this method to add the type. This will mean changing the signature of all of the reader methods in GenericDefaultReader and SpecificDefaultReader. This change would make some other things easier too.

Right now the ReflectDeserializer needs to be told the type of arrays and maps.
        /// <summary>
        /// C# type to create when deserializing and array. Must implement IList&lt;&gt;
        /// Default is System.Collections.Generic.List
        /// </summary>
        /// <value></value>
        public Type ListType { get => _listType; set => _listType = value; }
        /// <summary>
        /// C# type to create when deserializing a map. Must implement IDictionary&lt;,&gt; and the first
        /// type parameter must be a string. Default is System.Collections.Generic.Dictionary
        /// </summary>
        /// <value></value>
        public Type MapType { get => _mapType; set => _mapType = value; }

We could do the same for Generic fixed vs byte[]. This will mean that you cannot mix GenericFixed and byte[], and, since the class cache is static, the types the deserializer uses should be regarded as globals.

What do you think about making the bigger change and adding the type to all ReadXXXXXX methods?

Regarding IEnumerable<>
                The deserializer needs an Add method. IEnumerable<> doesn’t have one so unless it’s ok to have classes we can serialize but not deserialize then I think we need to stick with IList<>

From: Patrick Farry <Patrick.Farry@pb.com>
Date: Tuesday, June 11, 2019 at 8:52 AM
To: apache/avro <reply@reply.github.com>, apache/avro <avro@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: Re: [apache/avro] AVRO-2389: CSharp serializers for plain C# objects (#521)

Hi Brian,

Thanks for the reply. Responses follow:


  1.  .net framework build: Need to fix this but I don’t have easy access to a windows machine. The system doesn’t build cleanly on my Mac. I figured I would put the net40 targets back in when the pull request was closer to being final (or just revert the changes). I can do this now if you think it’s better.
  2.  Field numbers: I was trying to match the behavior of the codegen classes. I agree using a field name attribute would be better – I’ll make the change.
  3.  Reflect vs. POCO. Happy with either. I guess Reflect for consistency. Will make the change.
  4.  Fixed type to support GenericFixed. Yes, can do that.
  5.  IEnumerable vs. IList: Will try it out. I think the code only ever iterates through the collection so it should work.
  6.  WriteMap. Yes Will do. I hadn’t realized the implementation didn’t change much. Thanks.
  7.  WriteUnion. Same comment as for WriteMap.

I noticed the class caching isn’t going to work when the object has an array or map at the top level. Will fix this too and add tests.

From: Brian Lachniet <notifications@github.com>
Reply-To: apache/avro <reply@reply.github.com>
Date: Tuesday, June 11, 2019 at 4:21 AM
To: apache/avro <avro@noreply.github.com>
Cc: Patrick Farry <Patrick.Farry@pb.com>, Mention <mention@noreply.github.com>
Subject: Re: [apache/avro] AVRO-2389: CSharp serializers for plain C# objects (#521)

CAUTION: THIS EMAIL IS FROM AN EXTERNAL SOURCE. Internet links, office documents or other attachments may contain viruses. Do not click on a link, open or enable any file unless you trust the sender.

@blachniet requested changes on this pull request.

25 days later... Sorry my delay in getting back. I've got some initial review feedback. I'd still like to dive into this deeper and get a better understanding of the nuts and bolts, but I wanted to share what I've got so far.

________________________________

In lang/csharp/src/apache/main/Avro.main.csproj<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r284416378&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156389378&sdata=chkdEEK2sSZaGQpv3cG5%2BNTGfDHEXaSVG%2B7IJvnC%2Blg%3D&reserved=0>:

> @@ -19,7 +19,7 @@



   <Import Project="../../../common.props" />







   <PropertyGroup>



-    <TargetFrameworks>net40;netstandard2.0</TargetFrameworks>



+    <TargetFrameworks>netstandard2.0</TargetFrameworks>



Was there something in these changes that prevented you from targeting net40?

We are currently supporting .NET Framework 4.0+ in the Avro C# libraries. Removing it here would be a breaking change. I've been toying with the idea of dumping support for anything prior to 4.6.1 (.NET Standard 2.0 supports .NET Framework 4.6.1+<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Fdotnet%2Fstandard%2Fnet-standard%23net-implementation-support&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156399387&sdata=2L6YMRiQfP%2FPvCI9LnN%2B4zPbnsjngagc68AcoAJFMCg%3D&reserved=0>) but I have not actually gotten to the point of proposing that to the Avro team yet.

If we were to drop support for .NET 4.0, I think we would probably need to wait for a major release, e.g. 1.10.x. For now, I suggest that you add net40 back in to the target frameworks.
⬇️ Suggested change

-    <TargetFrameworks>netstandard2.0</TargetFrameworks>



+    <TargetFrameworks>net40;netstandard2.0</TargetFrameworks>



________________________________

In lang/csharp/src/apache/main/POCO/AvroAttribute.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292396937&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156409395&sdata=6o12xYGph19PMIMLjhnWtADFI3u4WDWElIrdIO4x8fY%3D&reserved=0>:

> +namespace Avro.POCO



+{



+    /// <summary>



+    /// Attribute to indicate how the class is mapped to the schema. If true the mapping is by position



+    /// and the class must have field attributes. If false the class is mapped by field name. Mapping by field



+    /// position allows you to use properties who's name doesnt match the schema.



+    ///



+    /// If there is no attribute mapping by name is the default.



+    /// </summary>



+    public class AvroAttribute : Attribute



+    {



+        /// <summary>



+        /// If true the class is mapped by position.



+        /// </summary>



+        /// <value></value>



+        public bool ByPosition { get; set; }



Why map properties to a field at a particular position? This seems fragile, especially since Avro schema evolution allows for the reordering of fields in a schema, which would change their "position".

In the documentation for this class, you mention that this would be used when your C# property name cannot match your Avro field name. This is definitely a scenario we want to handle. Maybe we should take an approach similar to that of the JsonPropertyAttribute<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.newtonsoft.com%2Fjson%2Fhelp%2Fhtml%2FJsonPropertyName.htm&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156409395&sdata=nnUxPFJcAFjGmL8waqyeycwCVxYZkBUe7wFKqoTBsgM%3D&reserved=0>. In the Avro approach, the AvroFieldAttribute would store the Avro field name to map a particular property to.

For example, the schema below defines a "release-date" field, which cannot be replicated exactly in a C# property name because of the hyphen.

{



  "type": "record",



  "name": "Movie",



  "fields" : [



    {"name": "title", "type": "string"},



    {"name": "release-date", "type": "long"}



  ]



}

I imagine we could map the C# property name to the appropriate Avro field name using the AvroFieldAttribute like so:

class Movie



{



    public string Title { get; set }







    [AvroFieldAttribute("release-date")]



    public long ReleaseDate { get; set; }



}

________________________________

In lang/csharp/src/apache/main/POCO/AvroAttribute.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292399459&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156419399&sdata=roQEmVD9UueL1qMO0Zy43iQP3zfqEOWwZ0Dwuf6jYqI%3D&reserved=0>:

> + * to you under the Apache License, Version 2.0 (the



+ * "License"); you may not use this file except in compliance



+ * with the License.  You may obtain a copy of the License at



+ *



+ *     http://www.apache.org/licenses/LICENSE-2.0



+ *



+ * Unless required by applicable law or agreed to in writing, software



+ * distributed under the License is distributed on an "AS IS" BASIS,



+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.



+ * See the License for the specific language governing permissions and



+ * limitations under the License.



+ */



+



+ using System;



+



+namespace Avro.POCO



The Java API has a similar feature called "Reflect". Should we stick with the "POCO" term to make it easier for C# devs to grok, or call it "Reflect" to match other Avro clients?

________________________________

In lang/csharp/src/apache/main/POCO/POCOWriter.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292400612&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156419399&sdata=kWMXnn7DjDJQxhI0OUANZmdpTZ%2FbNfRz7CFxgxR1Wa8%3D&reserved=0>:

> +        }



+



+        /// <summary>



+        /// Validates that the record is a fixed record object and that the schema in the object is the



+        /// same as the given writer schema. Writes the given fixed record into the given encoder



+        /// </summary>



+        /// <param name="schema">writer schema</param>



+        /// <param name="value">fixed object to write</param>



+        /// <param name="encoder">encoder to write to</param>



+        protected override void WriteFixed(FixedSchema schema, object value, Encoder encoder)



+        {



+            var fixedrec = value as byte[];



+            if (fixedrec == null)



+            {



+                throw new AvroTypeException("Fixed object is not derived from byte[]");



+            }



I think it would be nice to also allow derivatives of GenericFixed as value in this method. This would allow users of the library to use some of their auto-generated Specific API classes in the POCO API.

________________________________

In lang/csharp/src/apache/main/POCO/POCOWriter.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292401231&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156429412&sdata=Gn9RttvoJY76m4onVeiDc8Z7641M21oK8JKJu7pOs%2Fo%3D&reserved=0>:

> +            }



+



+            encoder.WriteFixed(fixedrec);



+        }



+



+        /// <summary>



+        /// Serialized an array. The default implementation calls EnsureArrayObject() to ascertain that the



+        /// given value is an array. It then calls GetArrayLength() and GetArrayElement()



+        /// to access the members of the array and then serialize them.



+        /// </summary>



+        /// <param name="schema">The ArraySchema for serialization</param>



+        /// <param name="value">The value being serialized</param>



+        /// <param name="encoder">The encoder for serialization</param>



+        protected override void WriteArray(ArraySchema schema, object value, Encoder encoder)



+        {



+            var arr = value as System.Collections.IList;



Could we allow IEnumerable<T> here rather than requiring an IList? DefaultSpecificWriter also requires an IList at this time, but maybe we could imporve on this here. Would allowing an IEnumerable cause problems?

________________________________

In lang/csharp/src/apache/main/POCO/POCOWriter.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292401952&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156439416&sdata=6FKfHTBnXt1BS1QMqT3QU%2BipRcaS4FSVcdGLowIK0go%3D&reserved=0>:

> +                encoder.StartItem();



+                Write(schema.ItemSchema, arr[i], encoder);



+            }



+



+            encoder.WriteArrayEnd();



+        }



+



+        /// <summary>



+        /// Writes the given map into the given encoder.



+        /// </summary>



+        /// <param name="schema">writer schema</param>



+        /// <param name="value">map to write</param>



+        /// <param name="encoder">encoder to write to</param>



+        protected override void WriteMap(MapSchema schema, object value, Encoder encoder)



+        {



+            if (value == null)



This is a very sensible check. We should probably add this to DefaultSpecificWriter as well (not necessarily as part of this PR). Actually, if not for this check, this WriteMap method is identical to the base implementation. After running this check, why not call base.WriteMap()?

________________________________

In lang/csharp/src/apache/main/POCO/POCOWriter.cs<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%23discussion_r292402258&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156439416&sdata=kONy9EQwJU1cuvmFICQm%2BcFhzsKa9R7oH1q3AyAQ%2FXQ%3D&reserved=0>:

> +            }



+



+            encoder.WriteMapEnd();



+        }



+



+        /// <summary>



+        /// Resolves the given value against the given UnionSchema and serializes the object against



+        /// the resolved schema member. The default implementation of this method uses



+        /// ResolveUnion to find the member schema within the UnionSchema.



+        /// </summary>



+        /// <param name="us">The UnionSchema to resolve against</param>



+        /// <param name="value">The value to be serialized</param>



+        /// <param name="encoder">The encoder for serialization</param>



+        protected override void WriteUnion(UnionSchema us, object value, Encoder encoder)



+        {



+            for (int i = 0; i < us.Count; i++)



WriteUnion appears identical to the base implementation. Do we really need to override it?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Favro%2Fpull%2F521%3Femail_source%3Dnotifications%26email_token%3DAIDVNMRN4OQ4KWJ63UTLBNTPZ6DFDA5CNFSM4HMLJ4Z2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBYX76PI%23pullrequestreview-238026557&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156449429&sdata=ezNR3l280LcFHMG%2F6G9TK97bHZDW%2BjULXcDg%2FF5ToqM%3D&reserved=0>, or mute the thread<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAIDVNMU6LNZYT23SG63NBSLPZ6DFDANCNFSM4HMLJ4ZQ&data=02%7C01%7CPatrick.Farry%40pb.com%7C72e86d59bb6548b01a2208d6ee5f01fa%7C8a4925a9fd8e4866b31cf719fb05dce6%7C1%7C0%7C636958489156449429&sdata=QXU6g0bp7q1KsPNhwxJ1L0BuatLT6uP80UTwiFnpkf8%3D&reserved=0>.

blachniet commented:
Hey @pa009fa,

As for the target frameworks, what you described makes sense. Let's wait until the PR is close to final, like you suggested.

Regarding GenericFixed: This seems like a fairly big change. I'd say that we should back off of this for now and go with your original approach.

Regarding IEnumerable<>: I agree, sounds like we need to stick with IList.

pa009fa commented:
Hi Brian,

I've done a bunch of stuff. 

Firstly cleaned up the tests files.

Changed the Avro.sln and added 2 new files so that running "dotnet test" from the solution folder identifies test projects and runs the tests. Without this change all projects attempt to run tests and the ones without tests throw errors. Note that ipc.test tests hangs on my machine - I'll try to find out why.

Made suggested changes to the Reflect files.

Got a solution for the GenericFixed. The code has the concept of Converters (borrowed from JSON.net) which are classes that convert between an Avro type and the type of the property. Main motivator was System.DateTime and System.DateTimeOffset, but they work for GenericFixed as well. See the TestFixed.cs test project.

Also got one for IEnumerable. The system still only works with a single collection type - i.e. cannot have both List<> and ConcurrentBag<> in the same Reader/Writer but I think it is a step towards the general functionality and should work for most use cases. Take a look at TestArray.cs.

Also, think it makes sense to wait until you make the .net standard build change. (P.S. dont think we need the MSBuild project anymore either). 

Thanks.

blachniet commented:
Hey @pa009fa,

Don't worry about **ipc.test**. That hasn't been working and we haven't been running it as part of our CI. In fact, I removed it from the solution in [AVRO-2422](https://github.com/apache/avro/pull/553). That's also were I configured all the projects to build with only Linux compatible targets (.NET Core and .NET Standard) unless building on Windows.

blachniet commented on lang/csharp/src/apache/main/Reflect/IAvroFieldConverter.cs:
```diff
@@ -0,0 +1,57 @@
+        /// Avro type
+        /// </summary>
+        /// <returns></returns>
+        Type GetAvroType();
```
```suggestion
        Type AvroType { get; }
```

pa009fa replied:
Following style in System.Reflection e.g. GetType()

OK to change if you think the property is better.

blachniet commented on lang/csharp/src/apache/main/Reflect/ClassCache.cs:
```diff
@@ -0,0 +1,209 @@
+            _defaultConverters.Add(converter);
+        }
+
+        public static void AddDefaultConverter<A,P>(Func<A,Schema, P> from, Func<P,Schema, A> to)
```
Add a comment to AddDefaultConverter, and any other externally visible elements. We haven't been good about this in the past in the C# project, but now's a great time to start, especially for new elements. 😄

pa009fa replied:
Done.

blachniet commented on lang/csharp/src/apache/test/Reflect/TestLogMessage.cs:
```diff
@@ -0,0 +1,123 @@
+            LogMessage deserialized = null;
+            using (var stream = new MemoryStream(serialized))
+            {
+                deserialized = avroReader.Read(default(LogMessage), new BinaryDecoder(stream));
```
We should probably assert that all the properties of `deserialized` match up with those of `logMessage`.

pa009fa replied:
Done

blachniet commented on lang/csharp/src/apache/test/Reflect/TestFromAvroProject.cs:
```diff
@@ -0,0 +1,252 @@
+                myArray3 = new List<List<object>>()
+            };
+
+            var zz = SerializeDeserialize(z);
```
We should assert that all the properties of `zz` equal those of `z`.

pa009fa replied:
Done

blachniet commented on lang/csharp/src/apache/test/Reflect/TestFromAvroProject.cs:
```diff
@@ -0,0 +1,252 @@
+                myArray3 = new List<List<object>>() { new List<object>() { 7.0, "def" } }
+            };
+
+            var zz = SerializeDeserialize(z);
```
We should assert that all the properties of `zz` equal those of `z`.

pa009fa replied:
Done

blachniet commented on lang/csharp/src/apache/main/Reflect/ReflectArrayHelper.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace Avro.Reflect
+{
+    public class ReflectArrayHelper
```
I'm reaching a bit here, but do you think that we could somehow indicate the type of array to create in the `AvroFieldAttribute`, rather than providing this array helper to the reader/writer, which means that all array-like structures must be represented using the same C# type?

pa009fa replied:
Great idea. Will add a new attribute for array types and allow the helper type to be specified there. Are you good with using dynamic in the standard helper - so anything with Add(), Clear() and Count() will work? There is so much dynamic code going on anyway..... The standard helper would cover everything implementing IList<> as well as ConcurrentBag<>.

pa009fa replied:
Got very close. Problem now is that Array is an Unnamed type so it isnt possible to look up the array type. I could switch it to a named type, with default name "array" like it is now. Alternatively is there some other field in the Avro spec we could use to bind the array to its C# type? What do you think?

pa009fa replied:
The spec says this: "Attributes not defined in this document are permitted as metadata, but must not affect the format of serialized data." What about adding a new attribute to array for the C# array helper binding?

pa009fa replied:
Added metadata attribute "helper"

pa009fa commented on lang/csharp/src/apache/test/Reflect/TestArray.cs:
```diff
@@ -0,0 +1,127 @@
+        [TestCase]
+        public void ListTest()
+        {
+            var schema = global::Avro.Schema.Parse(_simpleList);
```
Fixed.

pa009fa commented on lang/csharp/src/apache/test/Reflect/TestArray.cs:
```diff
@@ -0,0 +1,127 @@
+            public string S { get; set; }
+        }
+
+        private string _simpleList = @"
```
Done.

pa009fa replied:
Done

pa009fa commented on lang/csharp/src/apache/main/Reflect/IAvroFieldConverter.cs:
```diff
@@ -0,0 +1,57 @@
+        /// Property type
+        /// </summary>
+        /// <returns></returns>
+        Type GetPropertyType();
```
I'd originally had a property called "Type"... right, not smart. 

But I changed it to the methods. A lot of the reflection library uses methods like e.g. object.GetType() this so I was copying that style. I'm not really invested either way - let me know what you think.

pa009fa commented:
@blachniet review items are done except for:
- AvroType property vs. GetAvroType() method. Dont care too much but wanted to explain why I used the methods. Let me know what you want to do here.
- IEnumerable for Arrays - was going great but hit a snag. ArraySchema is an UnnamedSchema there is no name to use to determine which ArrayType to create and which array helper to create. Not sure of the best way forward: 
    - give up and just go with a single Array type, 
    - add some sort of extra info to Array in the schema that we can use to create the right ArrayHelper
    - add the type into all deserializer calls (changes the Generic and Specific deserializers but since they dont use the type it might not be too big a deal, but it is a lot of work).

pa009fa commented:
Ready for another review. Not sure why there is still a merge conflict.

blachniet commented:
@pa009fa, do you mind if I rebase this branch on master?

pa009fa commented:
Go ahead. Let me know what I need to do.

blachniet commented:
@pa009fa, I'm very sorry, but I accidentally pushed apache/avro master to pa009fa/avro master, rather than your rebased changes. I've created [pa009fa/avro PR 2](https://github.com/pa009fa/avro/pull/2) in your repository to get the rebased changes back into your master.

After those changes are back in your repo, we can reopen this PR. I'm very sorry for this screw up.

blachniet commented:
@pa009fa: Or, if you prefer, you could force-push your local copy of master to pa009fa/avro master, and you could then perform the rebase yourself. Let me know if I can be of any assistance. Sorry 😦

pa009fa commented:
@blachniet: I am a github noobie (and not that knowledgeable about git either). I'm trying to figure out how to rebase. All the changes are in 2 directories (main/Reflect and test/Reflect). I checked them in to master (which I think is a good thing because it squashed all of the changes). If it is easy for you to rebase then would you mind doing that part? I'll work on getting up to speed.

blachniet commented:
Thanks @pa009fa, I'll take a look at the new PR you created.

For future reference, this PR is superseded by #587.

PaluruSumanth commented:
@blachniet  Though we have the support for .net framework 4.6.1, it also has the dependency on net standard library, we see that target frameworks doesnt have .net framework 4.6.1, is that intended?
